crates/services/importer/src/importer.rs

```diff
@@ -0,0 +1,342 @@
+    transactional::StorageTransaction,
+    Error as StorageError,
+    IsNotFound,
+};
+use fuel_core_types::{
+    blockchain::{
+        consensus::{
+            Consensus,
+            Sealed,
+        },
+        primitives::{
+            BlockHeight,
+            BlockId,
+        },
+        SealedBlock,
+    },
+    fuel_vm::crypto::ephemeral_merkle_root,
+    services::{
+        block_importer::{
+            ImportResult,
+            UncommittedResult,
+        },
+        executor,
+        executor::{
+            ExecutionBlock,
+            ExecutionResult,
+        },
+        Uncommitted,
+    },
+};
+use std::sync::Arc;
+use tokio::sync::{
+    broadcast,
+    TryAcquireError,
+};
+
+#[cfg(test)]
+pub mod test;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("The commit is already in the progress: {0}.")]
+    SemaphoreError(TryAcquireError),
+    #[error("The wrong state of database during insertion of the genesis block.")]
+    InvalidUnderlyingDatabaseGenesisState,
+    #[error(
+        "The wrong state of database after execution of the block.\
+        The actual height is {1}, when the next expected height is {0}."
+    )]
+    InvalidDatabaseStateAfterExecution(BlockHeight, BlockHeight),
+    #[error("Got overflow during increasing the height.")]
+    Overflow,
+    #[error("The non-generic block can't have zero height.")]
+    ZeroNonGenericHeight,
+    #[error("The actual height is {1}, when the next expected height is {0}.")]
+    IncorrectBlockHeight(BlockHeight, BlockHeight),
+    #[error(
+        "Got another block id after validation of the block. Expected {0} != Actual {1}"
+    )]
+    BlockIdMismatch(BlockId, BlockId),
+    #[error("Some of the block fields are not valid: {0}.")]
+    FailedVerification(anyhow::Error),
+    #[error("The execution of the block failed: {0}.")]
+    FailedExecution(executor::Error),
+    #[error("It is not possible to skip transactions during importing of the block.")]
+    SkippedTransactionsNotEmpty,
+    #[error("It is not possible to execute the genesis block.")]
+    ExecuteGenesis,
+    #[error("The database already contains the data at the height {0}.")]
+    NotUnique(BlockHeight),
+    #[error(transparent)]
+    StorageError(#[from] StorageError),
+}
+
+#[cfg(test)]
+impl PartialEq for Error {
+    fn eq(&self, other: &Self) -> bool {
+        format!("{:?}", self) == format!("{:?}", other)
+    }
+}
+
+pub struct Importer<D, E, V> {
+    database: D,
+    executor: E,
+    verifier: V,
+    broadcast: broadcast::Sender<Arc<ImportResult>>,
+    guard: tokio::sync::Semaphore,
+}
+
+impl<D, E, V> Importer<D, E, V> {
+    pub fn new(config: Config, database: D, executor: E, verifier: V) -> Self {
+        let (broadcast, _) = broadcast::channel(config.max_block_notify_buffer);
+        Self {
+            database,
+            executor,
+            verifier,
+            broadcast,
+            guard: tokio::sync::Semaphore::new(1),
```

Comment:
Any reason to use semaphore instead of tokio mutex? It's basically the same thing but a little bit simpler https://docs.rs/tokio/latest/src/tokio/sync/mutex.rs.html#131

Reply:
Initially, it was `Mutex`, but Tommi suggested using `Semaphore`, and I agreed because there are two reasons:
- `Mutex` uses `Semaphore` under the hood
- With `Mutex`, we need to specify the type, and it looks like `Mutex<()>`, when with `Semaphore` we don't need that. Plus, `Mutex` is about access to resources, while `Semaphore` is about signaling mechanisms. Here we use it to signal that some process is already in progress.